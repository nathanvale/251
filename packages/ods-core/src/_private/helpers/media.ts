import {
  css,
  // FlattenInterpolation,
  // Interpolation,
  // SimpleInterpolation,
  ThemedStyledProps,
  DefaultTheme,
  // ThemedCssFunction,
  InterpolationFunction,
  SimpleInterpolation,
} from 'styled-components';

import {BreakpointVariants} from '@origin-digital/ods-themes';

export enum MediaLabels {
  xs = 'smaller',
  sm = 'phone',
  md = 'tablet',
  lg = 'desktop',
  xl = 'giant',
}

export enum MediaAliases {
  smaller = 'xs',
  phone = 'sm',
  tablet = 'md',
  desktop = 'lg',
  giant = 'xl',
}

export type Media = MediaLabels | MediaAliases;

// export type MediaTagFunction<
//   P extends {} = {},
//   T extends DefaultTheme = DefaultTheme
// > = (
//   strings: TemplateStringsArray | NonNullable<SimpleInterpolation>,
//   ...interpolations: Interpolation<ThemedStyledProps<P, T>>[]
// ) => FlattenInterpolation<ThemedStyledProps<P, T>>;

export type MapMediaToQuery = {[Key in Media]: InterpolationFunction<any>};

export type MediaObject = MapMediaToQuery & {
  min: MapMediaToQuery;
  max: MapMediaToQuery;
};

export type GetBreakpointsFn = <P extends {} = {}>(
  props: ThemedStyledProps<P, DefaultTheme>,
) => Record<BreakpointVariants, number>;

const defaultBreakpoints: Record<BreakpointVariants, number> = {
  xs: 575,
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
};

const getBreakpoints: GetBreakpointsFn = () => ({
  ...defaultBreakpoints,
});

export const media: MediaObject = (Object.keys(
  defaultBreakpoints,
) as Media[]).reduce(
  (accumulator, label) => {
    const minMedia = (
      strings: TemplateStringsArray,
      ...interpolations: SimpleInterpolation[]
    ) => css`
      @media (min-width: ${props => (getBreakpoints(props) as any)[label]}px) {
        ${css(strings, ...interpolations)}
      }
    `;

    const maxMedia = (
      strings: TemplateStringsArray,
      ...interpolations: SimpleInterpolation[]
    ) => css`
      @media (max-width: ${props => (getBreakpoints(props) as any)[label]}px) {
        ${css(strings, ...interpolations)}
      }
    `;

    accumulator[label] =
      label === 'xs' || label === 'smaller' ? maxMedia : minMedia;
    accumulator.max[label] = maxMedia;
    accumulator.min[label] = minMedia;

    return accumulator;
  },
  // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
  {min: {}, max: {}} as MediaObject,
);
