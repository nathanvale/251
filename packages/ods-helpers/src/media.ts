import {
  css,
  // FlattenInterpolation,
  // Interpolation,
  // SimpleInterpolation,
  ThemedStyledProps,
  // ThemedCssFunction,
  InterpolationFunction,
  SimpleInterpolation,
} from "styled-components";

import { BreakpointVariants } from "@origin-digital/ods-types";

enum MediaLabels {
  xs = "smaller",
  sm = "phone",
  md = "tablet",
  lg = "desktop",
  xl = "giant",
}

enum MediaAliases {
  smaller = "xs",
  phone = "sm",
  tablet = "md",
  desktop = "lg",
  giant = "xl",
}

type Media = MediaLabels | MediaAliases;

// export type MediaTagFunction<
//   P extends {} = {},
//   T extends {} = {}
// > = (
//   strings: TemplateStringsArray | NonNullable<SimpleInterpolation>,
//   ...interpolations: Interpolation<ThemedStyledProps<P, T>>[]
// ) => FlattenInterpolation<ThemedStyledProps<P, T>>;

type MapMediaToQuery = { [Key in Media]: InterpolationFunction<any> };

type MediaObject = MapMediaToQuery & {
  min: MapMediaToQuery;
  max: MapMediaToQuery;
};

type GetBreakpointsFn = <P extends {} = {}>(
  props: ThemedStyledProps<P, {}>
) => Record<BreakpointVariants, number>;

const defaultBreakpoints: Record<BreakpointVariants, number> = {
  xs: 575,
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
};

const getBreakpoints: GetBreakpointsFn = () => ({
  ...defaultBreakpoints,
});

export const media: MediaObject = (Object.keys(
  defaultBreakpoints
) as Media[]).reduce(
  (accumulator, label) => {
    const minMedia = (
      strings: TemplateStringsArray,
      ...interpolations: SimpleInterpolation[]
    ) => css`
      @media (min-width: ${(props) =>
          (getBreakpoints(props) as any)[label]}px) {
        ${css(strings, ...interpolations)}
      }
    `;

    const maxMedia = (
      strings: TemplateStringsArray,
      ...interpolations: SimpleInterpolation[]
    ) => css`
      @media (max-width: ${(props) =>
          (getBreakpoints(props) as any)[label]}px) {
        ${css(strings, ...interpolations)}
      }
    `;

    accumulator[label] =
      label === "xs" || label === "smaller" ? maxMedia : minMedia;
    accumulator.max[label] = maxMedia;
    accumulator.min[label] = minMedia;

    return accumulator;
  },
  { min: {}, max: {} } as MediaObject
);
