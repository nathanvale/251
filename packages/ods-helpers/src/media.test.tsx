import React from "react";
import { render } from "@testing-library/react";
import styled from "styled-components";
import { media } from "./media";

test("It should render with a xs media query", () => {
  const Test = styled.div`
    ${media.xs`
      color: red;
    `}
  `;
  const { container } = render(<Test />);
  expect(container.firstChild).toMatchSnapshot();
});

test("It should render with a sm media query", () => {
  const Test = styled.div`
    ${media.sm`
      color: red;
    `}
  `;
  const { container } = render(<Test />);
  expect(container.firstChild).toMatchSnapshot();
});

test("It should render with a md media query", () => {
  const Test = styled.div`
    ${media.md`
      color: red;
    `}
  `;
  const { container } = render(<Test />);
  expect(container.firstChild).toMatchSnapshot();
});

test("It should render with a lg media query", () => {
  const Test = styled.div`
    ${media.lg`
      color: red;
    `}
  `;
  const { container } = render(<Test />);
  expect(container.firstChild).toMatchSnapshot();
});

test("It should render with a xl media query", () => {
  const Test = styled.div`
    ${media.xl`
      color: red;
    `}
  `;
  const { container } = render(<Test />);
  expect(container.firstChild).toMatchSnapshot();
});
