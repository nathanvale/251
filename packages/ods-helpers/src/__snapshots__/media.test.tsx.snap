// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render with a lg media query 1`] = `
@media (min-width:992px) {
  .c0 {
    color: red;
  }
}

<div
  class="c0"
/>
`;

exports[`It should render with a md media query 1`] = `
@media (min-width:768px) {
  .c0 {
    color: red;
  }
}

<div
  class="c0"
/>
`;

exports[`It should render with a sm media query 1`] = `
@media (min-width:576px) {
  .c0 {
    color: red;
  }
}

<div
  class="c0"
/>
`;

exports[`It should render with a xl media query 1`] = `
@media (min-width:1200px) {
  .c0 {
    color: red;
  }
}

<div
  class="c0"
/>
`;

exports[`It should render with a xs media query 1`] = `
@media (max-width:575px) {
  .c0 {
    color: red;
  }
}

<div
  class="c0"
/>
`;
